### PR TITLE
Use Enumerable.Empty<> over Array.Empty<>

### PR DESCRIPTION
**Pull request by drewnoakes**
### Context

Observed while reading through code. I encountered this in a trace on a different code base and dug out my benchmarking from then.

Enumeration of `Enumerable.Empty<>` incurs zero allocations, whereas enumerating `Array.Empty<>` allocates 32 bytes. The former is also 40% faster.

|            Method |      Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Allocated |
|------------------ |----------:|----------:|----------:|------:|--------:|-------:|----------:|
|        ArrayEmpty | 34.285 ns | 0.7368 ns | 1.3839 ns |  1.00 |    0.00 | 0.0055 |      32 B |
|   EnumerableEmpty | 20.980 ns | 0.4713 ns | 0.6607 ns |  0.60 |    0.03 |      - |         - |

```c#
[SimpleJob(RuntimeMoniker.Net48)]
[MemoryDiagnoser]
public class EmptyEnumeratorBenchmarks
{
    [Benchmark(Baseline = true)] public bool ArrayEmpty() => Array.Empty<string>().GetEnumerator().MoveNext();

    [Benchmark] public bool EnumerableEmpty() => Enumerable.Empty<string>().GetEnumerator().MoveNext();
}
```

### Changes Made

Replace uses of `Array.Empty<>` with `Enumerable.Empty<>` where possible. Most of these changes appear related to build logging.

### Testing

CI.

### CODE REVIEW

**Comment by Therzok** [comment]
`foreach` over array/string is replaced with a for loop by the compiler. 

Array's GetEnumerator is an object.

It's possible Enumerable.Empty got inlined because it's an EmptyPartition.

https://source.dot.net/#System.Linq/System/Linq/Partition.SpeedOpt.cs,9611940f9dbc593c

**Review by rainersigwald — APPROVED** [review]

**Comment by Therzok** [comment]
Also, yeah, Enumerable.Empty, afaik, is an `IPartition<T>` which is the main type optimized for in LINQ.

**Comment by drewnoakes** [comment]
> `foreach` over array/string is replaced with a for loop by the compiler.

This is true when the declared type is an array. When the declared type is `IEnumerable` an enumerator is allocated.

**Comment by Therzok** [comment]
Yup, that's why I mentioned Array's GetEnumerator. I like this change.

File: src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs [hunk 1]
```diff
@@ -80,8 +81,8 @@ private ProjectLoggingContext(NodeLoggingContext nodeLoggingContext, int submiss
             ProjectPropertyInstanceEnumeratorProxy properties = null;
             ProjectItemInstanceEnumeratorProxy items = null;
 
-            IEnumerable<ProjectPropertyInstance> projectPropertiesEnumerator = projectProperties == null ? Array.Empty<ProjectPropertyInstance>() : null;
-            IEnumerable<ProjectItemInstance> projectItemsEnumerator = projectItems == null ? Array.Empty<ProjectItemInstance>() : null;
+            IEnumerable<ProjectPropertyInstance> projectPropertiesEnumerator = projectProperties == null ? Enumerable.Empty<ProjectPropertyInstance>() : null;
+            IEnumerable<ProjectItemInstance> projectItemsEnumerator = projectItems == null ? Enumerable.Empty<ProjectItemInstance>() : null;
```
Comment (Forgind) [inline src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs:85]: Better yet, inline these. It looks like there are several opportunities for simplifying this constructor.

**Review by Forgind — APPROVED** [review]

File: src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs:96]: Apologies, I didn't get to this before it was merged. Looks like it's been done in #7117 though.

Comment (Forgind) [reply src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs:96]: Yes! No worries; I saw that and marked this merge-when-branch-open.

**Pull request merged**